documentation/intermediate-ca.md

```diff
@@ -0,0 +1,199 @@
+
+See [PUP-3788](https://tickets.puppetlabs.com/browse/PUP-3788) for more information on chain revocation checking.
+
+### Generate an intermediate CA certificate request
+
+Save the following ruby script to a file, `csrgen.rb`. Make it executable (with `chmod u+x csrgen.rb`), and then run `./csrgen.rb`.
+
+    #!/opt/puppetlabs/puppet/bin/ruby
+    require 'puppet'
+    require 'openssl'
+
+    class Puppet::Application::GenCACSR < Puppet::Application
+      run_mode :master
+
+      def main
+        root_cert = OpenSSL::X509::Certificate.new(File.read(Puppet[:cacert]))
+        ca_key = OpenSSL::PKey::RSA.new(File.read(Puppet[:cakey]))
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        puts csr.to_s
+      end
+    end
+    Puppet::Application::GenCACSR.new.run
+
+Use this CSR to generate the intermediate CA certificate that Puppet will be using.
+
+### Submit the generated CSR for signing
+
+The signing procedure will depend on the PKI that your organization is deployed; consult your PKI administrators for more information.
+
+### Construct the CA certificate chain
+
+Puppet Server needs to present the full certificate chain to clients in order for the client to authenticate the server. The certificate chain is constructed by concatenating the CA certificates together, starting with the new intermediate CA certificate and descending to the root CA certificate.
+
+    cat puppet-intermediate-certificate.pem \
+        org-intermediate-certificate.pem \
+         org-root-certificate.pem > ca-bundle.pem
+
+In this example, `puppet-intermediate-certificate.pem` is signed CA certificate that you'll be using with your Puppet PKI, and will be given to you by your PKI administrators. The `org-intermediate-certificate.pem` and `org-root-certificate.pem` are the CA certificates in use by your organization; the exact number of CA certificates will depend on the structure of your organization's PKI.
+
+### Verify the Puppet master host certificate against the CA bundle
+
+Before installing the new certificate bundle, verify that the bundle can be used to verify the existing host certificate on the CA server. If this step fails then the CA certificate bundle is invalid and cannot be installed.
+
+    openssl verify -CAfile ca-bundle.pem $(puppet master --configprint hostcert)
+
+### Update the Puppet CA certificate bundle
+
+    cp ca-bundle.pem /etc/puppetlabs/puppet/ssl/ca/ca_crt.pem
```

Comment:
Would it be prudent to tell users to back up the existing CA certificate before doing this?